### ISSUE TITLE
Missing issuer attributes on authentication request.

### ISSUE THREAD

**Issue by SbGibson** [KentorIT/authservices#637]
Hi,

my IdP requires that, in the login request, the Issuer node have the Format and the NameQualifier attributes. 
Looking at to section 3.2.1 in SAML Core specification the Issuer is defined as a NameIDType that have those optional attributes.
Looking at the Saml2RequestBase code I saw that Issuer is defined as System.IdentityModel.Metadata.EntityId that have only the ID property. I thought to change from System.IdentityModel.Metadata.EntityId to System.IdentityModel.Tokens.Saml2NameIdentifier but I found 29 places where Issuer is used, also as key of the IdentityProviders dictionary so maybe I'm missing something. 
There's some other way to specify optional attributes of issuer without modifying all that code?

**Comment by AndersAbel** [comment]
You're right, all Issuer fields should be of `Saml2NameIdentifyer` type to conform with the spec. I obviously took a shortcut back when I started this project (never realized how complex the SAML2 specification is, nor how widespread the library would eventually be).

I think that such a change is reasonable (even though it would be breaking). With smart overloads of existing methods that wraps an `EntityId` in a `Saml2NameIdentifier` it should not break too many things for people. In most cases people are just *adding* stuff to the IdentityProviders dictionary and not actually reading.

What do you think @explunit?

**Comment by explunit** [comment]
I agree -- technically a breaking change but unlikely to break the most common usages.

**Comment by espositofulvio** [comment]
Hi,
I'm interested in this topic as well, is there any way with the current version to add those attributes when the XML is generated? Maybe inheriting from the Saml2AuthenticateRequest and rewriting the ToXElement() method? but how can I tell IdentityProvider to create my class then?

**Comment by AndersAbel** [comment]
@espositofulvio No, I don't think a workaround is possible. This need to be implemented in the library.
